core/src/test/java/org/apache/iceberg/avro/TestAvroNameMapping.java

```diff
@@ -0,0 +1,257 @@
+    Assert.assertNull("location field should not be read", projected.get("location_r5"));
+    Assert.assertEquals(34L, projected.get("id"));
+
+    // Table mapping partially project `location` map value
+    nameMapping = MappingUtil.create(new Schema(
+        Types.NestedField.required(0, "id", Types.LongType.get()),
+        Types.NestedField.optional(5, "location", Types.MapType.ofOptional(6, 7,
+            Types.StringType.get(),
+            Types.StructType.of(
+                Types.NestedField.required(1, "lat", Types.FloatType.get()))))));
+
+    projected = writeAndRead(writeSchema, readSchema, record, nameMapping);
+    Record projectedL1 = ((Map<String, Record>) projected.get("location")).get("l1");
+    Assert.assertNotNull("Field missing from table mapping is renamed", projectedL1.getSchema().getField("long_r2"));
+    Assert.assertNull("location.value.long, should not be read", projectedL1.get("long_r2"));
+  }
+
+  @Test
+  public void testMissingRequiredFields() {
+    Schema writeSchema = new Schema(
+        Types.NestedField.required(19, "x", Types.IntegerType.get()),
+        Types.NestedField.optional(18, "y", Types.IntegerType.get()));
+
+    Record record = new Record(AvroSchemaUtil.convert(writeSchema, "table"));
+    record.put("x", 1);
+    record.put("y", 2);
+
+    // table mapping not projecting a required field 'x'
+    NameMapping nameMapping = MappingUtil.create(new Schema(
+        Types.NestedField.optional(18, "y", Types.IntegerType.get())));
+
+
+    Schema readSchema = writeSchema;
+    AssertHelpers.assertThrows("Missing required field in nameMapping",
+        IllegalArgumentException.class, "Missing required field: x",
+        // In this case, pruneColumns result is an empty record
+        () -> writeAndRead(writeSchema, readSchema, record, nameMapping));
+  }
+
+  @Test
+  public void testArrayProjections() throws Exception {
+    Schema writeSchema = new Schema(
+        Types.NestedField.required(0, "id", Types.LongType.get()),
+        Types.NestedField.optional(22, "point",
+            Types.ListType.ofOptional(21, Types.StructType.of(
+                Types.NestedField.required(19, "x", Types.IntegerType.get()),
+                Types.NestedField.optional(18, "y", Types.IntegerType.get())
+            ))
+        )
+    );
+
+    Record record = new Record(AvroSchemaUtil.convert(writeSchema, "table"));
+    record.put("id", 34L);
+    Record pointRecord = new Record(AvroSchemaUtil.fromOption(
+        AvroSchemaUtil.fromOption(record.getSchema().getField("point").schema()).getElementType()));
+    pointRecord.put("x", 1);
+    pointRecord.put("y", 2);
+    record.put("point", ImmutableList.of(pointRecord));
+
+    NameMapping nameMapping = MappingUtil.create(new Schema(
+        // Optional array field missing.
+        Types.NestedField.required(0, "id", Types.LongType.get())));
+
+    Schema readSchema = new Schema(
```

Comment:
Isn't this the write schema?

Reply:
Yes, in this test, read and write schema matches

Reply:
Can you update this to `readSchema = writeSchema` like the test above?